### PR TITLE
K8s: Add integration test infra, and fix authz patterns

### PR DESCRIPTION
**Pull request by ryantxu**
This is pulled out from https://github.com/grafana/grafana/pull/75709, and fixes the authz issues discovered while adding tests 🎉 

In particular -- the `union` authz helper will return a final answer when *any* authorizer says Allow or Deny -- this means that our authorizers should only return Allow when they know it is absolutely OK (eg, writing to the correct org is necessary, but not sufficient!)

### CODE REVIEW

**Review by DanCech — APPROVED** [review]
LGTM

File: pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go
Comment (toddtreece) [inline pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go:31]: why wouldn't we say allow here?

**Review by toddtreece — COMMENTED** [review]
falling back to allow feels risky, but the changes to `org_id` and `stack_id` make sense to me 👍. i'm not sure i understand the `org_role` change, but the addition of integration tests is 💯

File: pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go
Comment (ryantxu) [reply pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go:31]: the problem is that this clobers every other decision -- so being an editor does not mean you can edit everything, it only means you have the option to edit (more FGAC can/will have more explicit options)

Yes the fallback to allow all is odd, but I think the alternative is using something other than `union` as our base.  Perhaps we want one that checks if anything says Deny, then if anything says Allow?  (but that would not work with `NewPrivilegedGroups(user.SystemPrivilegedGroup)`)

Comment (ryantxu) [reply pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go:31]: note the other change is to return Deny -- when the role is unkown

File: pkg/tests/apis/types_discoverty.go
Comment (mildwonkey) [inline pkg/tests/apis/types_discoverty.go:2]: I think this filename should be `types_discovery.go`

File: pkg/services/grafana-apiserver/auth/authorizer/provider.go
Comment (mildwonkey) [inline pkg/services/grafana-apiserver/auth/authorizer/provider.go:35]: Do you expect to change this behavior eventually? It seems overly permissive. If it's just temporary and not meant to be long-term please add a TODO/FIXME comment to help us find it later?

File: pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go
Comment (toddtreece) [reply pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go:31]: > being an editor does not mean you can edit everything, it only means you have the option to edit
ah ok, that makes sense

>  FGAC can/will have more explicit options
👍 so it sounds like we could make that the final decision once it's available?

**Review by toddtreece — APPROVED** [review]

File: pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go
Comment (ryantxu) [reply pkg/services/grafana-apiserver/auth/authorizer/org/org_role.go:31]: Yes -- where FGAC has rules, it can say yes|no and be the final decision, but when it does not this will still be the fallback path

File: pkg/services/grafana-apiserver/auth/authorizer/provider.go
Comment (ryantxu) [reply pkg/services/grafana-apiserver/auth/authorizer/provider.go:35]: so no -- not a fixme.  The the alternative could be combining the logic from org_id|stack_id + org_role into one authorizer and saying that if nothing is Denied, then the request is OK.  I don't have any strong opinions if that makes it more or less clear

File: pkg/tests/apis/types_discoverty.go
Comment (ryantxu) [reply pkg/tests/apis/types_discoverty.go:2]: omg -- what is the "t" for 🤔

File: pkg/services/grafana-apiserver/auth/authorizer/provider.go
Comment (mildwonkey) [reply pkg/services/grafana-apiserver/auth/authorizer/provider.go:35]: I follow that. I still don't love it, but what's the current behavior? I'd expect that we always fall back to deny if there isn't an explicit allow, and we should implement the same behavior? I also get that EVERYTHING IS DIFFERENT) in this model - don't worry, I am not going to block this PR on this conversation.

**Comment by ryantxu** [comment]
@mildwonkey -- i removed the typo file 🤣 -- that was added as the first step to creating a dynamic client, but @toddtreece has shown me an client-go approach that should be able to do the same thing.  I'll add that in a new PR.

I think the authz logic here is correct, however we *may* want to change the approach of depending nothing from:
org_id|stack_id + org_role saying "no opinion" to mean yes, and combine that into a single authorizer that will return yes

**Comment by ryantxu** [comment]
After merging this, i'll add a branch to evaluate a combined authz authorizer

**Pull request merged**
